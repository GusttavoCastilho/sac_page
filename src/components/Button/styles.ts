import styled, { css, DefaultTheme } from "styled-components";
import { ButtonProps } from ".";
import { darken } from "polished";

type WrapperProps = { hasIcon: boolean } & Pick<
  ButtonProps,
  "color" | "minimal" | "iconSize"
>;

const WrapperModifiers = {
  withIcon: (theme: DefaultTheme, iconSize: number) => css`
    img {
      width: ${iconSize}px;
      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,
  minimal: () => css`
    border-bottom-left-radius: 10px;
    border-bottom-right-radius: 10px;
  `,
};

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, color, hasIcon, minimal, iconSize }) => css`
    height: 4rem;

    background-color: ${theme.colors[color!]};
    color: ${theme.colors.white};

    padding: ${theme.spacings.xxsmall} ${theme.spacings.small};

    border-radius: ${minimal ? "0px" : "28px"};
    border: none;

    cursor: pointer;

    display: flex;
    align-items: center;

    text-transform: uppercase;

    font-size: ${theme.font.sizes.xsmall};
    font-weight: ${theme.font.bold};

    :hover {
      background-color: ${darken(0.05, theme.colors[color!])};
    }

    ${!!hasIcon && WrapperModifiers.withIcon(theme, iconSize!)};
    ${!!minimal && WrapperModifiers.minimal()};
  `}
`;
