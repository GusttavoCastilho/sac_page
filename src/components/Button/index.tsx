import React, { ButtonHTMLAttributes } from "react";
import * as S from "./styles";

export type ButtonProps = {
  children: React.ReactNode;
  color?: "yellow" | "purple_dark" | "blue";
  icon?: string;
  iconSize?: number;
  minimal?: boolean;
} & ButtonHTMLAttributes<HTMLButtonElement>;

const Button = ({
  children,
  color = "yellow",
  icon,
  iconSize,
  minimal = false,
  ...props
}: ButtonProps) => (
  <S.Wrapper
    color={color}
    hasIcon={!!icon}
    minimal={minimal}
    iconSize={iconSize}
    {...props}
  >
    {icon && <img src={icon} alt="icon" />}
    <span>{children}</span>
  </S.Wrapper>
);

export default Button;
