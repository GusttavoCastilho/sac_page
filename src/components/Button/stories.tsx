import React from "react";
import { Story, Meta } from "@storybook/react/types-6-0";
import Button from ".";

export default {
  title: "Button",
  component: Button,
  argTypes: {
    children: {
      type: "string",
    },
    icon: {
      type: "",
    },
  },
} as Meta;

export const Default: Story = (args) => (
  <Button {...args}>Seja um franqueado</Button>
);

export const WithIcon: Story = (args) => (
  <Button {...args}>Seja um franqueado</Button>
);

WithIcon.args = {
  icon: "/icons/editar.png",
};

export const WithMinimal: Story = (args) => (
  <Button {...args} minimal>
    Seja um franqueado
  </Button>
);
