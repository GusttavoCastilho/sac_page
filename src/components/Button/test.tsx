import { screen } from "@testing-library/react";
import { renderWithTheme } from "../../utils/helpers";
import Button from ".";

describe("<Button />", () => {
  it("should render a correctly struct button and styles", () => {
    const { container } = renderWithTheme(<Button>Seja um franqueado</Button>);

    expect(
      screen.getByRole("button", { name: /Seja um franqueado/i })
    ).toHaveStyle({
      height: "4rem",
      padding: "0.8rem 2.4rem",
      "font-size": "1.2rem",
      "text-transform": "uppercase",
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render an icon version", () => {
    renderWithTheme(
      <Button icon="/icons/editar.png">Seja um franqueado</Button>
    );

    expect(screen.getByText(/Seja um franqueado/i)).toBeInTheDocument();
  });

  it("should render a minimal version", () => {
    renderWithTheme(
      <Button icon="/icons/editar.png" minimal>
        Seja um franqueado
      </Button>
    );

    expect(
      screen.getByRole("button", { name: /Seja um franqueado/i })
    ).toHaveStyle({
      "border-bottom-left-radius": "10px",
      "border-bottom-right-radius": "10px",
    });
  });
});
