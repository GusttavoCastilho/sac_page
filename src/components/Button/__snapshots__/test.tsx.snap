// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a correctly struct button and styles 1`] = `
.c0 {
  height: 4rem;
  background-color: #E1A607;
  color: #FFFFFF;
  padding: 0.8rem 2.4rem;
  border-radius: 28px;
  border: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-transform: uppercase;
  font-size: 1.2rem;
  font-weight: 600;
}

.c0:hover {
  background-color: #c89406;
}

<button
  class="c0"
  color="yellow"
>
  <span>
    Seja um franqueado
  </span>
</button>
`;
